src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java

```diff
@@ -392,11 +443,33 @@ public void close(Collection<TopicPartition> partitions) {
               topicPartition.partition());
           partitionsToChannel.remove(partitionChannelKey);
         });
+  }
+
+  private void closeInParallel(Collection<TopicPartition> partitions) {
+    CompletableFuture<?>[] futures =
+        partitions.stream().map(this::closeTopicPartition).toArray(CompletableFuture[]::new);
+
+    CompletableFuture.allOf(futures).join();
+  }
+
+  private CompletableFuture<Void> closeTopicPartition(TopicPartition topicPartition) {
+    String key = partitionChannelKey(topicPartition.topic(), topicPartition.partition());
+
+    TopicPartitionChannel topicPartitionChannel = partitionsToChannel.get(key);
+
     LOGGER.info(
-        "Closing {} partitions and remaining partitions which are not closed are:{}, with size:{}",
-        partitions.size(),
-        partitionsToChannel.keySet().toString(),
-        partitionsToChannel.size());
+        "Closing partitionChannel:{}, partition:{}, topic:{}",
+        topicPartitionChannel == null ? null : topicPartitionChannel.getChannelNameFormatV1(),
+        topicPartition.topic(),
+        topicPartition.partition());
+
+    // It's possible that some partitions can be unassigned before their respective channels are
+    // even created.
+    return topicPartitionChannel == null
+        ? CompletableFuture.completedFuture(null)
```

Comment:
thinking out loud, does this need a NPE handling? or we are fine in `closeInParallel`?

Reply:
No, null handling isn't needed. 
1. The returned future carries a `Void` value, so only `null` can be passed in there.
2. We don't do anything with the values returned by futures. The only thing we care about is their completion.

Reply:
perhaps these futures could return "key" instead - then you could simply pass method pointer to the thenAccept(partitonsToChannel::remove) ?

Reply:
`TopicPartitionChannel` which returns a future from `closeChannelAsync` knows nothing about a "key"`. Therefore, it can't return it.